common/model/transfer.c

```diff
@@ -1,676 +1,474 @@
+  }
+  memcpy(value_out, output, sizeof(transfer_value_out_t));
+
+  tf->type = VALUE_OUT;
+  tf->address = address;
+  tf->value = value;
+  tf->tag = tag;
+  tf->timestamp = timestamp;
+  tf->meta = (void*)value_out;
+  return tf;
+}
+
+// create a value_int transfer object.
+transfer_t* transfer_value_in_new(flex_trit_t const* const address,
+                                  flex_trit_t const* const tag, int64_t value,
+                                  flex_trit_t const* const data,
+                                  size_t data_len, uint64_t timestamp) {
+  transfer_value_in_t* value_in = NULL;
+  transfer_t* tf = NULL;
+  if (!address) {
+    log_error(TRANSFER_LOGGER_ID, "[%s:%d] address cannot be NULL.\n", __func__,
+              __LINE__);
+    return NULL;
+  }
+
+  if (!tag) {
+    log_error(TRANSFER_LOGGER_ID, "[%s:%d] tag cannot be NULL.\n", __func__,
+              __LINE__);
+    return NULL;
+  }
+
+  if (!data && data_len != 0) {
+    log_error(TRANSFER_LOGGER_ID, "[%s:%d] data length should be 0.\n",
+              __func__, __LINE__);
+    return NULL;
+  }
+
+  tf = (transfer_t*)calloc(1, sizeof(transfer_t));
+  if (!tf) {
+    log_error(TRANSFER_LOGGER_ID, "[%s:%d] Out of Memory.\n", __func__,
+              __LINE__);
+    return NULL;
+  }
+
+  value_in = (transfer_value_in_t*)calloc(1, sizeof(transfer_value_in_t));
+  if (!value_in) {
+    free(tf);
+    log_error(TRANSFER_LOGGER_ID, "[%s:%d] Out of Memory.\n", __func__,
+              __LINE__);
+    return NULL;
+  }
+  value_in->len = data_len;
+  value_in->data = data;
+
+  tf->type = VALUE_IN;
+  tf->address = address;
+  tf->tag = tag;
+  tf->value = value;
+  tf->timestamp = timestamp;
+  tf->meta = (void*)value_in;
+  return tf;
+}
+
+void transfer_free(transfer_t** transfer) {
+  if (transfer || *transfer) {
+    free((*transfer)->meta);
+    free(*transfer);
+    *transfer = NULL;
+  }
+}
+
+size_t transfer_transactions_count(transfer_t* tf) {
+  transfer_data_t* data = NULL;
+  transfer_value_out_t* value_out = NULL;
+  switch (tf->type) {
+    case DATA:
+      data = (transfer_data_t*)tf->meta;
+      return (data->len + NUM_TRITS_SIGNATURE - 1) / NUM_TRITS_SIGNATURE;
+    case VALUE_OUT:
+      // security level
+      value_out = (transfer_value_out_t*)tf->meta;
+      return value_out->security;
+    case VALUE_IN:
+      // fixed
+      return 1;
+  }
+  return 0;
+}
+
+transfer_ctx_t* transfer_ctx_new() {
+  transfer_ctx_t* tf_ctx = (transfer_ctx_t*)calloc(1, sizeof(transfer_ctx_t));
+  if (!tf_ctx) {
+    log_error(TRANSFER_LOGGER_ID, "[%s:%d] Out of Memory.\n", __func__,
+              __LINE__);
+    return NULL;
+  }
+  memset(tf_ctx->bundle, FLEX_TRIT_NULL_VALUE, FLEX_TRIT_SIZE_243);
+  return tf_ctx;
+}
+
+void transfer_ctx_free(transfer_ctx_t* transfer_ctx) { free(transfer_ctx); }
+
+bool transfer_ctx_init(transfer_ctx_t* transfer_ctx, transfer_t* transfers[],
+                       size_t len) {
+  size_t i;
+  int64_t total = 0;
```

Comment:
Maybe set `transfer_ctx->count = 0` before the loop in case we re-init it ?

Reply:
should not re-init a init method.

Reply:
Maybe put a comment then ?